File: packages/cortex-js/tests/resources/backends.test.ts
import { describe, it, expect } from 'vitest';
import { ValidationError } from '../../src/errors.js';
import { mockFetch, createClient } from '../helpers.js';

describe('backends', () => {
  describe('list', () => {
    it('lists all backends', async () => {
      const mockResponse = {
        backends: [{ id: 'b-1', name: 'OpenAI', url: 'https://api.openai.com', createdAt: '2024-01-01', updatedAt: '2024-01-01' }],
      };
      const fetch = mockFetch(mockResponse);
      const client = createClient(fetch);

      const result = await client.backends.list();

      expect(result.backends).toHaveLength(1);
      expect(result.backends[0]!.name).toBe('OpenAI');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/backends');
      expect(opts.method).toBe('GET');
    });
  });

  describe('create', () => {
    it('registers a backend', async () => {
      const mockResponse = { id: 'b-2', name: 'Anthropic', url: 'https://api.anthropic.com', createdAt: '2024-01-01', updatedAt: '2024-01-01' };
      const fetch = mockFetch(mockResponse);
      const client = createClient(fetch);

      const result = await client.backends.create({ name: 'Anthropic', url: 'https://api.anthropic.com' });

      expect(result.name).toBe('Anthropic');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/backends');
      expect(opts.method).toBe('POST');
    });

    it('throws ValidationError when name is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.backends.create({ name: '', url: 'https://x.com' })).rejects.toThrow(ValidationError);
    });

    it('throws ValidationError when url is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.backends.create({ name: 'Test', url: '' })).rejects.toThrow(ValidationError);
    });
  });

  describe('update', () => {
    it('updates a backend', async () => {
      const fetch = mockFetch({ id: 'b-1', name: 'Updated', url: 'https://x.com', createdAt: '2024-01-01', updatedAt: '2024-01-02' });
      const client = createClient(fetch);

      const result = await client.backends.update('b-1', { name: 'Updated' });

      expect(result.name).toBe('Updated');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/backends/b-1');
      expect(opts.method).toBe('PATCH');
    });

    it('throws ValidationError when id is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.backends.update('', { name: 'x' })).rejects.toThrow(ValidationError);
    });
  });

  describe('delete', () => {
    it('removes a backend', async () => {
      const fetch = mockFetch({ id: 'b-1', deleted: true });
      const client = createClient(fetch);

      const result = await client.backends.delete('b-1');

      expect(result.deleted).toBe(true);

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/backends/b-1');
      expect(opts.method).toBe('DELETE');
    });

    it('throws ValidationError when id is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.backends.delete('')).rejects.toThrow(ValidationError);
    });
  });

  describe('discover', () => {
    it('discovers models on a backend', async () => {
      const fetch = mockFetch({ models: [{ id: 'm-1', name: 'gpt-4' }] });
      const client = createClient(fetch);

      const result = await client.backends.discover('b-1');

      expect(result.models).toHaveLength(1);
      expect(result.models[0]!.name).toBe('gpt-4');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/backends/b-1/discover');
      expect(opts.method).toBe('POST');
    });

    it('throws ValidationError when id is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.backends.discover('')).rejects.toThrow(ValidationError);
    });
  });

  describe('updateModel', () => {
    it('updates a model display name', async () => {
      const fetch = mockFetch({ id: 'm-1', name: 'gpt-4', displayName: 'GPT-4 Turbo' });
      const client = createClient(fetch);

      const result = await client.backends.updateModel('b-1', 'm-1', { displayName: 'GPT-4 Turbo' });

      expect(result.displayName).toBe('GPT-4 Turbo');

      const [url, opts] = fetch.mock.calls[0] as [string, RequestInit];
      expect(url).toBe('https://admin.test.com/admin/backends/b-1/models/m-1');
      expect(opts.method).toBe('PATCH');
    });

    it('throws ValidationError when backendId is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.backends.updateModel('', 'm-1', { displayName: 'x' })).rejects.toThrow(ValidationError);
    });

    it('throws ValidationError when modelId is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.backends.updateModel('b-1', '', { displayName: 'x' })).rejects.toThrow(ValidationError);
    });

    it('throws ValidationError when displayName is missing', async () => {
      const client = createClient(mockFetch({}));
      await expect(client.backends.updateModel('b-1', 'm-1', { displayName: '' })).rejects.toThrow(ValidationError);
    });
  });
});

File: packages/cortex-dart/test/resources/backends_test.dart
import 'dart:convert';

import 'package:nfinitmonkeys_cortex_sdk/cortex_sdk.dart';
import 'package:http/http.dart' as http;
import 'package:test/test.dart';

import '../helpers.dart';

Map<String, dynamic> sampleBackend({
  String id = 'backend-123',
  String name = 'OpenAI Backend',
}) =>
    {
      'id': id,
      'name': name,
      'url': 'https://api.openai.com',
      'status': 'active',
      'provider': 'openai',
      'models': [
        {'id': 'gpt-4', 'displayName': 'GPT-4'},
      ],
      'createdAt': '2024-01-01T00:00:00Z',
    };

void main() {
  group('BackendsResource', () {
    late CortexClient client;

    tearDown(() => client.close());

    test('list returns backends', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {
          'data': [
            sampleBackend(id: 'b-1', name: 'Backend A'),
            sampleBackend(id: 'b-2', name: 'Backend B'),
          ],
        }),
      );

      final backends = await client.backends.list();
      expect(backends, hasLength(2));
      expect(backends.first.id, 'b-1');
      expect(backends.last.name, 'Backend B');
    });

    test('list handles empty response', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {'data': <dynamic>[]}),
      );

      final backends = await client.backends.list();
      expect(backends, isEmpty);
    });

    test('create sends correct body', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, sampleBackend(), requests),
      );

      final backend = await client.backends.create(
        name: 'New Backend',
        url: 'https://api.example.com',
        provider: 'custom',
      );

      expect(backend.id, 'backend-123');

      final body = jsonDecode(requests.first.body) as Map<String, dynamic>;
      expect(body['name'], 'New Backend');
      expect(body['url'], 'https://api.example.com');
      expect(body['provider'], 'custom');
    });

    test('create validates empty name', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.backends.create(name: '', url: 'https://example.com'),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('create validates empty url', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.backends.create(name: 'Test', url: ''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('update sends PATCH request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, sampleBackend(), requests),
      );

      final backend = await client.backends.update(
        id: 'backend-123',
        name: 'Updated Backend',
      );

      expect(backend.id, 'backend-123');
      expect(requests.first.method, 'PATCH');
    });

    test('update validates empty id', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.backends.update(id: ''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('delete sends DELETE request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, {}, requests),
      );

      await client.backends.delete('backend-123');

      expect(requests.first.method, 'DELETE');
      expect(
        requests.first.url.path,
        contains('/admin/backends/backend-123'),
      );
    });

    test('discover sends POST request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, sampleBackend(), requests),
      );

      final backend = await client.backends.discover('backend-123');

      expect(backend.id, 'backend-123');
      expect(requests.first.method, 'POST');
      expect(
        requests.first.url.path,
        contains('/admin/backends/backend-123/discover'),
      );
    });

    test('discover validates empty id', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.backends.discover(''),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('updateModel sends PATCH request', () async {
      final requests = <http.Request>[];
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: recordingMockClient(200, {
          'id': 'gpt-4',
          'displayName': 'GPT-4 Turbo',
        }, requests),
      );

      final model = await client.backends.updateModel(
        backendId: 'backend-123',
        modelId: 'gpt-4',
        displayName: 'GPT-4 Turbo',
      );

      expect(model.id, 'gpt-4');
      expect(model.displayName, 'GPT-4 Turbo');
      expect(requests.first.method, 'PATCH');
      expect(
        requests.first.url.path,
        contains('/admin/backends/backend-123/models/gpt-4'),
      );
    });

    test('updateModel validates empty backendId', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.backends.updateModel(
          backendId: '',
          modelId: 'gpt-4',
          displayName: 'GPT-4',
        ),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('updateModel validates empty displayName', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        httpClient: mockClient(200, {}),
      );

      expect(
        () => client.backends.updateModel(
          backendId: 'b-1',
          modelId: 'gpt-4',
          displayName: '',
        ),
        throwsA(isA<CortexValidationException>()),
      );
    });

    test('handles 401 errors', () async {
      client = CortexClient(
        apiKey: 'sk-cortex-test-key-1234',
        maxRetries: 0,
        httpClient: mockClient(401, 'Unauthorized'),
      );

      expect(
        () => client.backends.list(),
        throwsA(isA<CortexAuthenticationException>()),
      );
    });
  });
}
